### DISCUSSION TITLE
Pytest with Microdot

### DISCUSSION THREAD

**Discussion by maxi07** [miguelgrinberg/microdot#157] [Q&A] [answered]
Hi there,
I was wondering if there are existing examples of how to implement (async) pytests with Microdot?

**Comment by miguelgrinberg** [comment]
The unit tests in this repository use pytest. Maybe that can give you some ideas? https://github.com/miguelgrinberg/microdot/blob/main/tests/test_microdot_asyncio.py

**Reply by maxi07** [reply]
I tried following, but i can only see the use of UnitTests. Do you have a simple example on how to use Microdot with pytest?

**Reply by maxi07** [reply]
this is my current apporach:

```python
import pytest

@pytest.fixture
def client(stub):
    stub.apply({
        'rp2': '[mock]',
        'neopixel': '[mock]',
        'machine': '[mock]',
        'network': '[mock]'
    })
    import settings
    return settings.app

def test_get_networks(client, mocker, stub):
    stub.apply({
        'rp2': '[mock]',
        'neopixel': '[mock]',
        'machine': '[mock]',
        'network': '[mock]'
    })
    # Mock the functions or methods used in the route
    mocker.patch('modules.wifi_setup.get_available_networks', return_value=[{"ssid": "TestSSID", "rssi": -93, "channel": 11}, {"ssid": "TestSSID2", "rssi": -53, "channel": 13}])

    response = client.get('/get-networks')

    assert response.status_code == 200
    assert response.content_type == 'application/json'
    data = json.loads(response.data)
    assert data == [{"ssid": "TestSSID", "rssi": -93, "channel": 11}, {"ssid": "TestSSID2", "rssi": -53, "channel": 13}]
```

In the _settings.py_, I run
```python
app = Microdot()
```

My Microdot routes are defined in the _routes.py_
```python
@settings.app.get("/get-networks")
async def get_networks(request):
    log_info("Received get networks request")
    networks = remove_duplicates(get_available_networks())
    json_data = json.dumps(networks)
    response = Response(json_data)
    response.headers['Content-Type'] = 'application/json'
    return response
```

When running the test, pytest failes with
`FAILED tests/test_webserver.py::test_get_networks - AttributeError: 'function' object has no attribute 'status_code'`, which seems like that the response object is _None_

**Reply by miguelgrinberg** [reply, marked as answer]
When I said that this repository uses pytest, I meant that it uses pytest as test runner. I choose to structure my tests in classes, I do not like fixtures. This is still all running under pytest, which has full support for tests given in classes.

You aren't showing enough code for me to tell for sure, but it seems you are writing your tests using sync Python, but your application uses asyncio. That does not work, the test client for microdot_asyncio is async too. Look at the tests in the file I linked above, they are all wrapped in the `_run()` helper function that puts the call inside a loop and awaits it.

**Reply by maxi07** [reply]
sync testing with async webserver was the issue. Thanks!
